org.eclipse.xtext/src/org/eclipse/xtext/workspace/FileProjectConfig.java

```diff
@@ -0,0 +1,118 @@
+/**
+ * Copyright (c) 2015, 2017 itemis AG (http://www.itemis.eu) and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.eclipse.xtext.workspace;
+
+import static org.eclipse.xtext.xbase.lib.IterableExtensions.*;
+
+import java.io.File;
+import java.util.HashSet;
+import java.util.Set;
+
+import org.eclipse.emf.common.util.URI;
+import org.eclipse.xtext.util.UriUtil;
+
+import com.google.common.base.Objects;
+
+public class FileProjectConfig implements IProjectConfig {
+	private final URI path;
+
+	private final String name;
+
+	private final Set<FileSourceFolder> sourceFolders = new HashSet<>();
+
+	private final IWorkspaceConfig workspaceConfig;
+
+	public FileProjectConfig(URI path) {
+		this(path, (IWorkspaceConfig) null);
+	}
+
+	public FileProjectConfig(URI path, String name) {
+		this(path, name, (IWorkspaceConfig) null);
+	}
+
+	public FileProjectConfig(File root, String name) {
+		this(root, name, (IWorkspaceConfig) null);
+	}
+
+	public FileProjectConfig(final File root) {
+		this(root, ((IWorkspaceConfig) null));
+	}
+
+	public FileProjectConfig(final URI path, final IWorkspaceConfig workspaceConfig) {
+		this(path, path.lastSegment(), workspaceConfig);
+	}
+
+	public FileProjectConfig(final URI path, final String name, final IWorkspaceConfig workspaceConfig) {
+		this.path = UriUtil.toFolderURI(path);
+		this.name = name;
+		this.workspaceConfig = (workspaceConfig != null) ? workspaceConfig : new WorkspaceConfig(this);
+	}
+
+	public FileProjectConfig(final File root, final String name, final IWorkspaceConfig workspaceConfig) {
+		this(UriUtil.createFolderURI(root), name, workspaceConfig);
+	}
+
+	public FileProjectConfig(final File root, final IWorkspaceConfig workspaceConfig) {
+		this(UriUtil.createFolderURI(root), root.getName(), workspaceConfig);
+	}
+
+	public FileSourceFolder addSourceFolder(final String relativePath) {
+		FileSourceFolder sourceFolder = new FileSourceFolder(this, relativePath);
+		this.sourceFolders.add(sourceFolder);
+		return sourceFolder;
+	}
+
+	@Override
+	public FileSourceFolder findSourceFolderContaining(final URI member) {
+		return findFirst(this.sourceFolders, f -> f.contains(member));
+	}
+
+	@Override
+	public boolean equals(final Object obj) {
+		if ((obj instanceof FileProjectConfig)) {
+			return Objects.equal(this.path, ((FileProjectConfig) obj).path);
+		}
+		return false;
+	}
+
+	@Override
+	public int hashCode() {
+		return this.path.hashCode();
+	}
+
+	@Override
+	public String toString() {
+		StringBuilder builder = new StringBuilder()
```

Comment:
String concatenation is easier to read. The compiler will create the boiler plate for us.

Reply:
done